src/Chapter3.hs

```diff
@@ -1005,7 +1134,113 @@ properties using typeclasses, but they are different data types in the end.
 Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
-
+newtype Potion = Potion Int
+newtype Spell = Spell Int
+
+class Fighter a where
+  attack :: a -> a -> a
```

Comment:
Answering your question, you can change the type of the other fighter and add the `Figther` constraint to it. With this approach, you know all information about `b` to work with it in generic way and reusing all functions that `Fighter` has. Nothing stops you from reusing the same `Figther` constraint on other arguments of its methods :slightly_smiling_face: 
```suggestion
  attack :: Fighter b => a -> b -> a
```

Reply:
Thanks @chshersh! Thanks a lot for your help. I think I got it this time!